contracts/ExoticPositionalMarkets/ExoticPositionalMarketManager.sol

```diff
@@ -0,0 +1,739 @@
+        emit MarketResolved(_marketAddress, _outcomePosition);
+    }
+
+    function cancelMarket(address _marketAddress) external {
+        require(isActiveMarket(_marketAddress), "Market is not active");
+        require(
+            msg.sender == oracleCouncilAddress || msg.sender == owner || msg.sender == creatorAddress[_marketAddress],
+            "Invalid address"
+        );
+        if (msg.sender != owner) {
+            require(oracleCouncilAddress != address(0), "Invalid address");
+        }
+        // Creator can cancel if it is the only ticket holder or only one that placed open bid
+        if (msg.sender == creatorAddress[_marketAddress]) {
+            require(
+                IExoticPositionalMarket(_marketAddress).canCreatorCancelMarket(),
+                "Market can not be cancelled by creator"
+            );
+            cancelledByCreator[_marketAddress] = true;
+        }
+        if (IExoticPositionalMarket(_marketAddress).paused()) {
+            require(msg.sender == owner, "only pDAO");
+        }
+        IExoticPositionalMarket(_marketAddress).cancelMarket();
+        resolverAddress[msg.sender] = safeBoxAddress;
+        if (cancelledByCreator[_marketAddress]) {
+            IExoticPositionalMarket(_marketAddress).claimWinningTicketOnBehalf(creatorAddress[_marketAddress]);
+        }
+        removeActiveMarket(_marketAddress);
+        emit MarketCanceled(_marketAddress);
+    }
+
+    function resetMarket(address _marketAddress) external onlyOracleCouncilAndOwner {
+        require(isActiveMarket(_marketAddress), "Market not active");
+        if (IExoticPositionalMarket(_marketAddress).paused()) {
+            require(msg.sender == owner, "only pDAO");
+        }
+        IExoticPositionalMarket(_marketAddress).resetMarket();
+        emit MarketReset(_marketAddress);
+    }
+
+    function sendRewardToDisputor(
+        address _market,
+        address _disputorAddress,
+        uint _amount
+    ) external onlyOracleCouncilAndOwner {
+        IExoticRewards(exoticRewards).sendRewardToDisputoraddress(_market, _disputorAddress, _amount);
+        // emit RewardSentToDisputorForMarket(_market, _disputorAddress, _amount);
+    }
+
+    function issueBondsBackToCreatorAndResolver(address _marketAddress) external nonReentrant {
+        require(
+            IExoticPositionalMarket(_marketAddress).canUsersClaim() || cancelledByCreator[_marketAddress],
+            "Not claimable"
+        );
+        require(
+            IThalesBonds(thalesBonds).getCreatorBondForMarket(_marketAddress) > 0 ||
+                IThalesBonds(thalesBonds).getResolverBondForMarket(_marketAddress) > 0,
+            "Bonds already claimed"
+        );
+        IThalesBonds(thalesBonds).issueBondsBackToCreatorAndResolver(_marketAddress);
+    }
+
+    function disputeMarket(address _marketAddress, address _disputor) external onlyOracleCouncil {
```

Comment:
add a check in all methods that have a market passed to them "isKnownMarket". Maintain a list of known markets as AddressSetLib (check PositionalMarketManager). Reason being that we cant know 100% a market was created through the manager without this check.

Reply:
refactored